add detection for custom libraries registered by ld.so.conf

PeterMocary opened the pull request:
The in-place upgrade doesn't support custom libraries linked using the ld.so configuration. The new actor introduced in this PR detects if the configuration was tempered with and creates high severity report in such case. In order to detect customization in the ld.so configuration, it needs to check the main `/etc/ld.so.conf` and the `/etc/ld.so.conf.d/`. 

Firstly the `/etc/ld.so.conf` where one can include other configuration files (by default `/etc/ld.so.conf.d/*.conf` is included for packages to copy their configuration in) and specify other directories for third parry libraries as well. The actor expects only includes in this file since the default configuration seems to be using the `/etc/ld.so.conf.d` directory to extend the configuration rather then adding libraries into the main config (based on the default configuration on rhel7 and rhel8).

Secondly, the included configs need configs need to be evaluated. The actor flags a config file as custom whenever it doesn't belong to a package or when the package is not Red Hat signed. This way only configuration that is supported by Red Hat will not be detected as custom.

Manually tested on internal Vagrant box rhel7 and rhel8. I left some of the debug logs in there for easy testing, please just mark those that shouldn't stay in your review and I'll remove them.

Jira ref.: OAMG-4460
BZ ref.: [BZ-1927700](https://bugzilla.redhat.com/show_bug.cgi?id=1927700) / [RHEL-11958](https://issues.redhat.com/browse/RHEL-11958)

PeterMocary committed:
add detection for custom libraries registered by ld.so.conf

The in-place upgrade process does not support custom libraries
and also does not handle customized configuration of dynamic linked.
In such a case it can happen (and it happens) that the upgrade could
break in critical phases when linked libraries dissapear or are not
compatible with the new system.

We cannot decide whether or not such a custom configuration affects
the upgrade negatively, so let's detect any customisations
or unexpected configurations related to dynamic linker and in such
a case generate a high severity report, informing user about the
possible impact on the upgrade process.

Currently it's detectect:
  * modified default LD configuration: /etc/ld.so.conf
  * drop int configuration files under /etc/ld.so.conf.d/ that are
    not owned by any RHEL RPMs
  * envars: LD_LIBRARY_PATH, LD_PRELOAD

Jira ref.: OAMG-4460 / RHEL-11958
BZ ref.: BZ 1927700

PeterMocary commented on repos/system_upgrade/common/actors/checkldconf/libraries/checkldsoconfiguration.py:
```diff
@@ -0,0 +1,98 @@
+    ])
+
+
+def _is_included_ld_so_config_custom(config_path):
```
I will maybe look into the `rpm -V` used for verification and maybe incorporate it in. This is a reasonable concern we should maybe address.

PeterMocary replied:
I further enhanced the detection of custom configuration files when they belong to a RH-signed RPM by running a `rpm -Va` command to detect any form of modification. Please let me know what you think about this approach.

matejmatuska replied:
I think this approach is better, even though it might produce false positives if, for example, a comment was added to a config file. Such change would get flagged as customization even though it doesn't functionally change the config. Since the produced report isn't a inhibitor anymore I think it's okay to not consider such an edge case. Also to truly say whether the config change was functional we would have to parse diffs which is IMHO an overkill.

matejmatuska replied:
@pirat89 What do you think about this?

pirat89 replied:
Tsorry for the late response. yes, we should check whether we speak about configuration files made by RH signed packages or not as we expected in case of RH signed packages (if configs are not modified) that these are ok. However, we need to check the behaviour in case of RH SCL.

PeterMocary commented on repos/system_upgrade/common/actors/checkldconf/libraries/checkldsoconfiguration.py:
```diff
@@ -0,0 +1,98 @@
+LIST_FORMAT_PREFIX = '\n    - '
+
+
+def _read_file(file_path):
+    with open(file_path, 'r') as fd:
+        return fd.readlines()
```
Talked about this offline and decided to keep this due to the fact that open is tricky to mock. If somebody has better or alternative way, please share.

PeterMocary commented on repos/system_upgrade/common/actors/checkldconf/libraries/checkldsoconfiguration.py:
```diff
@@ -0,0 +1,98 @@
+LD_SO_CONF_DIR = '/etc/ld.so.conf.d'
+LD_SO_CONF_MAIN = '/etc/ld.so.conf'
+LD_SO_CONF_DEFAULT_INCLUDE = 'ld.so.conf.d/*.conf'
+LIST_FORMAT_PREFIX = '\n    - '
```
Changed.

PeterMocary commented on repos/system_upgrade/common/actors/checkldconf/libraries/checkldsoconfiguration.py:
```diff
@@ -0,0 +1,98 @@
+            cfg_glob = os.path.join('/etc', cfg_glob) if not os.path.isabs(cfg_glob) else cfg_glob
+            included_configs.append(cfg_glob)
+        elif line:
+            other_lines.append(line)
```
I added comment detection in both main config file and included config files. The implementation now considers also the fact that the included config might contain comments only, The configuration file is not considered custom in that case, as it has no effect on configuration.

PeterMocary commented on repos/system_upgrade/common/actors/checkldconf/libraries/checkldsoconfiguration.py:
```diff
@@ -0,0 +1,98 @@
+        is_custom = not has_package(InstalledRedHatSignedRPM, package_name)
+    except CalledProcessError:
+        is_custom = True
+    api.current_logger().debug('The following config file is{}considered a custom '
```
Removed.

PeterMocary commented:
The actor now creates High severity report only. The inhibitor was removed as discussed in the BZ thread.

dkubek reviewed (dismissed):
LGTM! Did not find any major flaw in the code. Manually tested on a VM for both RHEL7/8. In testing I tried combinations of empty files, comments and custom paths for both ``/etc/ld.so.conf`` and custom drop-ins in ``/etc/ld.so.conf.d`` and the report is created as advertised.

matejmatuska commented on repos/system_upgrade/common/actors/checkldsoconf/libraries/checkldsoconfiguration.py:
```diff
@@ -0,0 +1,127 @@
+    is_modified = False
+    try:
+        run(['rpm', '-Va', package_name])
+    except CalledProcessError as err:
```
+1

PeterMocary replied:
Good point! I modified the function.

matejmatuska commented on repos/system_upgrade/common/actors/checkldsoconf/libraries/checkldsoconfiguration.py:
```diff
@@ -0,0 +1,127 @@
+def _is_modified(config_path, package_name):
+    """ Decide if the configuration file was modified based on the package it belongs to. """
+    is_modified = False
+    result = run(['rpm', '-Va', package_name], checked=False)
```
Is there a reason why you use `rpm -Va`? Although this does work I think a better approach would be to use `rpm -V <package name>` or, even better, `rpm -Vf <file name>` which would save you some of the parsing. See the package selection options in `man rpm`.

matejmatuska replied:
Also when I try `rpm -Va` with the full kernel package name instead of `kernel-core` it doesn't detect a modified `/etc/ld.so.conf.d/kernel-4.18.0-477.10.1.el8_8.x86_64.conf` whereas both `rpm -V` and `rpm -Vf` do.

PeterMocary replied:
Ou yes, `-Vf` seems like the way to go, I just can't read and somehow mismatched those...

PeterMocary replied:
Fixed thanks!

matejmatuska reviewed (dismissed):
Apart from the comments the code works as expected on RHEL8

PeterMocary commented on repos/system_upgrade/common/actors/checkldsoconf/libraries/checkldsoconfiguration.py:
```diff
@@ -0,0 +1,127 @@
+        line = line.strip()
+        if line.startswith('include'):
+            cfg_glob = line.split(' ', 1)[1].strip()
+            cfg_glob = os.path.join('/etc', cfg_glob) if not os.path.isabs(cfg_glob) else cfg_glob
+            included_configs.append(cfg_glob)
```
Fixed thanks!

pirat89 commented:
@PeterMocary fix please unit-tests and rebase

PeterMocary commented:
The failing tests weren't caused by this PR, I made them green by disabling some not really important warnings ([logging-not-lazy](https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/logging-not-lazy.html), [logging-format-interpolation](https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/logging-format-interpolation.html)) and fixing the [deprecated-method](https://pylint.pycqa.org/en/stable/user_guide/messages/warning/deprecated-method.html) warning for `current_logger.warn()` method. Due to the fact that I decided to disable some warnings I will not squash it yet and wait for your feedback @pirat89.

pirat89 commented:
@PeterMocary  seems good to me. switch the last commit for the linters to be the first one and keep it separated.

pirat89 commented on repos/system_upgrade/common/actors/checkldsoconf/libraries/checkldsoconfiguration.py:
```diff
@@ -0,0 +1,123 @@
+
+def _is_modified(config_path):
+    """ Decide if the configuration file was modified based on the package it belongs to. """
+    result = run(['rpm', '-Vf', config_path], checked=False)
+    if result['exit_code'] == 0:
+        return False
+    output = result['stdout']
+    # Note that the condition is relaxed so that files different only
+    # in the modification timestamp won't be considered as modified.
+    modification_flags = output.split(' ', 1)[0]
+    return modification_flags.strip('.') != 'T'
```
@PeterMocary  Reading the comment, it seems to me the only problem you are trying to resolve by the check of the output is whether the file contains some changes (hash, chmod, ...) or whether the timestamp has been changed only. correct? if so, this could simplify the solution:

```suggestion
    result = run(['rpm', '-Vf', '--nomtime', config_path], checked=False)
    if result['exit_code'] == 0:
        return False
    return True
```
This will ignore change of timestamp and report any other changes.

PeterMocary replied:
I changed this to check only the checksum of the file as we discussed offline.

pirat89 reviewed (commented):
this should be tested with an installed and enabled RH SCL to see the behaviour. I am not sure in that one particular case whether such configs are tracked by installed RH RPMs. 

Also The summary should be extended. Current summary does not include information why we report this to customer. what are the reasons? what could happen? etc. The title is speaking about third party libraries but in the summary I cannot see anything like that.

Also be aware of speaking about unsupported configuration files. We should rather say that these files seems to not be managed by RH products and as such they are not treated during the upgrade. The responsibility for them is on users and potentially they could impact the in-place upgrade negatively.

pirat89 commented on repos/system_upgrade/common/actors/checkldsoconf/libraries/checkldsoconfiguration.py:
```diff
@@ -0,0 +1,123 @@
+def _report_custom_ld_so_configuration(summary):
+    reporting.create_report([
+        reporting.Title(
+            'Third party libraries linked with ld.so.conf are not supported.'
```
@oamg/developers  any other ideas?

PeterMocary replied:
I think that the first one fits well.

pirat89 reviewed (changes requested):
see my previous comments. we can sync later to discuss more details.

PeterMocary commented:
I addressed the comments and rebased.
 - added detection of `LD_LIBRARY_PATH` as discussed offline.
 - RHSCL won't be solved in this PR (bit more information in the ticket OAMG-4460)
 - New report looks something like this:
```
----------------------------------------
Risk Factor: high
Title: Detected customized configuration for dynamic linker.
Summary: Custom configurations to the dynamic linker could potentially impact the upgrade in a negative way. The custom configuration includes modifications to /etc/ld.so.conf, custom or modified drop in config files in the /etc/ld.so.conf.d directory and additional entries in the LD_LIBRARY_PATH variable. These modifications configure the dynamic linker to use different libraries that might not be provided by Red Hat products. The following custom configurations were detected by leapp:
- The /etc/ld.so.conf file has unexpected contents:
    - aaa
- The following drop in config files were marked as custom:
    - /etc/ld.so.conf.d/custom_config.conf
    - /etc/ld.so.conf.d/dyninst-x86_64.conf
- The variable "LD_LIBRARY_PATH" contains unexpected dynamic linker configuration.
Remediation: [hint] Remove or revert the custom ld.so configuration and apply the changes using the ldconfig command.
Key: cc9bd972af70b7a27f66a37b11a00dcfcb73b1bc
----------------------------------------
```
Feel free to suggest changes if you think that this is not informative enough.

PeterMocary commented on repos/system_upgrade/common/actors/checkldsoconf/libraries/checkldsoconfiguration.py:
```diff
@@ -0,0 +1,123 @@
+    is_default_include_present = '/etc/' + LD_SO_CONF_DEFAULT_INCLUDE in included_configs
+    if not is_default_include_present:
+        api.current_logger().debug('The default include "' + LD_SO_CONF_DEFAULT_INCLUDE +
+                                   '" is not present in the ' + LD_SO_CONF_MAIN + ' file.')
```
fixed

pirat89 commented on repos/system_upgrade/common/actors/checkldsoconf/libraries/checkldsoconfiguration.py:
```diff
@@ -0,0 +1,133 @@
+FMT_LIST_SEPARATOR_2 = '\n    - '
+
+
+def _read_file(file_path):
+    with open(file_path, 'r') as fd:
+        return fd.readlines()
+
+
```
@PeterMocary  the actor is in the checksphase - where we allow to work only with data obtained via msgs. any  interaction with the system is prohibited here (r/w files, shell cmds execution - run, etc.). In this case we need an actor that produce msg with all important data, and this actor will consume such a msg and create a report based on it.

PeterMocary replied:
Separated into two actors.

pirat89 commented on repos/system_upgrade/common/actors/checkldsoconf/libraries/checkldsoconfiguration.py:
```diff
@@ -0,0 +1,133 @@
+                                                 FMT_LIST_SEPARATOR_2, FMT_LIST_SEPARATOR_2.join(custom_ld_configs))
+
+    if os.environ.get(LD_LIBRARY_PATH_VAR, None):
+        custom_configurations += ('{}The variable "{}" contains unexpected dynamic linker '
+                                  'configuration.').format(FMT_LIST_SEPARATOR_1, LD_LIBRARY_PATH_VAR)
```
```suggestion
        custom_configurations += (
            '{}The "{}" variable contains unexpected dynamic linker configuration.'
            .format(FMT_LIST_SEPARATOR_1, LD_LIBRARY_PATH_VAR)
        )
```

PeterMocary replied:
Changed :+1:

pirat89 commented on repos/system_upgrade/common/actors/checkldsoconf/libraries/checkldsoconfiguration.py:
```diff
@@ -0,0 +1,133 @@
+                   'way. The custom configuration includes modifications to {}, custom or modified drop in config '
+                   'files in the {} directory and additional entries in the {} variable. These modifications '
+                   'configure the dynamic linker to use different libraries that might not be provided by Red Hat '
+                   'products. The following custom configurations were detected '
```
```suggestion
                   'products or might not be present during the whole upgrade process. '
                   'The following custom configurations were detected '
```

PeterMocary replied:
Added.

pirat89 commented on repos/system_upgrade/common/actors/checkldsoconf/libraries/checkldsoconfiguration.py:
```diff
@@ -0,0 +1,133 @@
+            'Detected customized configuration for dynamic linker.'
+        ),
+        reporting.Summary(summary),
+        reporting.Remediation(hint=('Remove or revert the custom ld.so configuration and apply the changes '
+                                    'using the ldconfig command.')),
```
I am sorry for broken formatting. /o\
IIRC, active SCLs are messing with the envar and in such a case we cannot call it custom configuration. From this point, suggesting that in case of active SCLs they should disable them, could resolve one of detected problems. We do not need to check here if there are some SCLs really if we just say that in such a case it could be source of detected problems.

```suggestion
        reporting.Remediation(hint=('Remove or revert the custom dynamic linker configurations and apply the changes '
                                    'using the ldconfig command. In case of possible active software collections we suggest disabling them persistently.')),
```

PeterMocary replied:
I see.. added.

pirat89 reviewed (changes requested):
I am sorry I get to the proper review so late. I completely overlooked earlier that the actor in checksphase is interacting with the system, this needs to be split into two actors. The report looks good, just with minor change(s).

pirat89 commented:
@PeterMocary  btw, deadline has been shift to Thu - see my msg on slack with more details.

PeterMocary commented:
I rebased the branch and addressed the comments. The original actor was split into two actors: 

- scan_dynamic_linker_configuration - contains most of the logic and creates new model named DynamicLinkerConfiguration
- check_dynamic_linker_configuration - checks if the configuration is custom based on the message produced by above actor and prints a report

I also added LD_PRELOAD variable to the mix, since it allows user to set some libraries as well

The new full report looks like this:
```
----------------------------------------
Risk Factor: high
Title: Detected customized configuration for dynamic linker.
Summary: Custom configurations to the dynamic linker could potentially impact the upgrade in a negative way. The custom configuration includes modifications to /etc/ld.so.conf, custom or modified drop in config files in the /etc/ld.so.conf.d directory and additional entries in the LD_LIBRARY_PATH or LD_PRELOAD variables. These modifications configure the dynamic linker to use different libraries that might not be provided by Red Hat products or might not be present during the whole upgrade process. The following custom configurations were detected by leapp:
- The /etc/ld.so.conf file has unexpected contents:
    - custom line 123
    - custom other line
- The following drop in config files were marked as custom:
    - /etc/ld.so.conf.d/mariadb-x86_64.conf
    - /etc/ld.so.conf.d/custom.conf
- The following variables contain unexpected dynamic linker configuration:
    - LD_LIBRARY_PATH
    - LD_PRELOAD
Remediation: [hint] Remove or revert the custom dynamic linker configurations and apply the changes using the ldconfig command. In case of possible active software collections we suggest disabling them persistently.
Key: cc9bd972af70b7a27f66a37b11a00dcfcb73b1bc
----------------------------------------
```

I tested it only once to generate this report on RHEL7.

PeterMocary commented:
@pirat89 @matejmatuska @dkubek When you have time please review, so I can fix possible problems before the deadline :)

pirat89 commented on repos/system_upgrade/common/models/dynamiclinker.py:
```diff
@@ -0,0 +1,23 @@
+from leapp.topics import SystemFactsTopic
+
+
+class ConfigFile(Model):
+    topic = SystemFactsTopic
+
+    path = fields.String()
+    modified = fields.Boolean()
+
+
+class MainConfigFile(ConfigFile):
+    topic = SystemFactsTopic
+
+    modified_lines = fields.List(fields.String(), default=[])
```
@PeterMocary These models names are generic too much. Even when it is not expected to produce them directly, I am afraid that it could cause some problems in future - e.g. if someone else creates model `ConfigFile`, the behaviour will be biased when importing classes to actors.. What about to use rather e.g. `LDConfigFile` and `LDMainConfigFile` instead? Also please add docstrings to classes and fields in case of models. To make it explicit when particular classes are not expected to be used as "messages" but only part of a specific msg, and make clear what each field is supposed to represent. We will use it in future for automatic generation of documentation for messages.

PeterMocary replied:
Ugh wanted to add some comments but it totally slipped my mind. The naming concerns were addressed as well. :+1:

pirat89 commented:
@PeterMocary by a quick look, i found problematic just names of config classes in models, but rest of the code seems good to me. I will get to it tomorrow (wed) to finish it. thanks for the fast changes!

PeterMocary commented:
Tests are failing on unrelated error now. Apparently, pylint decided that ``target_user_space_creator`` is too large. ``[C0302 too-many-lines] Too many lines in module (1047/1000) File: repos/system_upgrade/common/actors/targetuserspacecreator/libraries/userspacegen.py, line 1, in ``

pirat89 commented:
I did rebase (with squash) and updated the commit msg to be more descriptive. Tests should be passing now. waiting for the results

dkubek reviewed (commented):
Had time to only look through code and apart from minor sugestions lgtm.

pirat89 commented on repos/system_upgrade/common/actors/scandynamiclinkerconfiguration/actor.py:
```diff
@@ -0,0 +1,23 @@
+
+class ScanDynamicLinkerConfiguration(Actor):
+    """
+    Scan the dynamoc linker configuration and find modifications.
```
interesting that it has not been caught by the spellchecker. I will fix it. good eyes!

pirat89 commented on repos/system_upgrade/common/actors/checkdynamiclinkerconfiguration/libraries/checkdynamiclinkerconfiguration.py:
```diff
@@ -0,0 +1,70 @@
+        )
+
+    if custom_configurations:
+        summary = (
+            'Custom configurations to the dynamic linker could potentially impact the upgrade in a negative '
+            'way. The custom configuration includes modifications to {}, custom or modified drop in config '
+            'files in the {} directory and additional entries in the {} or {} variables. These modifications '
+            'configure the dynamic linker to use different libraries that might not be provided by Red Hat '
+            'products or might not be present during the whole upgrade process. The following custom '
+            'configurations were detected by leapp:{}'
+            .format(LD_SO_CONF_MAIN, LD_SO_CONF_DIR, LD_LIBRARY_PATH_VAR, LD_PRELOAD_VAR, custom_configurations)
+        )
```
I've been thinking about that and you persuaded me that it will be still better to do it. I am going on that

pirat89 replied:
done.

pirat89 reviewed (approved):
lgtm and tests passed.

pirat89 commented:
Tested manually, it seems everything is ok.

- with installed and active SCL + custom config added
```
----------------------------------------
Risk Factor: high
Title: Detected customized configuration for dynamic linker.
Summary: Custom configurations to the dynamic linker could potentially impact the upgrade in a negative way. The custom configuration includes modifications to /etc/ld.so.conf, custom or modified drop in config files in the /etc/ld.so.conf.d directory and additional entries in the LD_LIBRARY_PATH or LD_PRELOAD variables. These modifications configure the dynamic linker to use different libraries that might not be provided by Red Hat products or might not be present during the whole upgrade process. The following custom configurations were detected by leapp:
- The following drop in config files were marked as custom:
    - /etc/ld.so.conf.d/myconf.conf
- The following variables contain unexpected dynamic linker configuration:
    - LD_LIBRARY_PATH
Remediation: [hint] Remove or revert the custom dynamic linker configurations and apply the changes using the ldconfig command. In case of possible active software collections we suggest disabling them persistently.
Key: cc9bd972af70b7a27f66a37b11a00dcfcb73b1bc
----------------------------------------
```

- same as above but mariadb ld conf has been modified
----------------------------------------
Risk Factor: high
Title: Detected customized configuration for dynamic linker.
Summary: Custom configurations to the dynamic linker could potentially impact the upgrade in a negative way. The custom configuration includes modifications to /etc/ld.so.conf, custom or modified drop in config files in the /etc/ld.so.conf.d directory and additional entries in the LD_LIBRARY_PATH or LD_PRELOAD variables. These modifications configure the dynamic linker to use different libraries that might not be provided by Red Hat products or might not be present during the whole upgrade process. The following custom configurations were detected by leapp:
- The following drop in config files were marked as custom:
    - /etc/ld.so.conf.d/mariadb-x86_64.conf
    - /etc/ld.so.conf.d/myconf.conf
- The following variables contain unexpected dynamic linker configuration:
    - LD_LIBRARY_PATH
Remediation: [hint] Remove or revert the custom dynamic linker configurations and apply the changes using the ldconfig command. In case of possible active software collections we suggest disabling them persistently.
Key: cc9bd972af70b7a27f66a37b11a00dcfcb73b1bc
----------------------------------------
